File: JAVA_code/Lecture_28.java
                                        //Lecture-28
                                        //july-26

import java.util.*;
public class Lecture_28
{   

    //1. To get min-cost for 2D array traversal  (only two moves allowed)

    public static int minCost(int[][] local,int[][] global)                         
    {
        for(int r=local.length-1;r>=0;r--)
        {
            for(int c=local[0].length-1;c>=0;c--)
            {
                if(r==local.length-1 && c==local[0].length-1)
                {   
                    global[r][c]=local[r][c];
                }
                else if(r==local.length-1)
                {
                    global[r][c]=local[r][c]+global[r][c+1];
                }
                else if(c==local[0].length-1)
                {
                    global[r][c]=global[r+1][c]+local[r][c];
                }
                else
                {
                    global[r][c]=Math.min(global[r][c+1],global[r+1][c])+local[r][c];
                }       
            }
        }

        return global[0][0];
    }

    //REVIEWED.








    //2. To get min-cost Path. (using dp to get min-path, and directed recursion to print paths.)

    public static void minCost2(int[][] local,int[][] global)       //DP function to get min-cost.
    {
        for(int r=local.length-1;r>=0;r--)
        {
            for(int c=local[0].length-1;c>=0;c--)
            {
                if(r==local.length-1 && c==local[0].length-1)
                {   
                    global[r][c]=local[r][c];
                }
                else if(r==local.length-1)
                {
                    global[r][c]=local[r][c]+global[r][c+1];
                }
                else if(c==local[0].length-1)
                {
                    global[r][c]=global[r+1][c]+local[r][c];
                }
                else
                {
                    global[r][c]=Math.min(global[r][c+1],global[r+1][c])+local[r][c];
                }       
            }
        }

    }



    public static void minPath(int[][] global2,int i,int j,String ans)        //directed recusion function.
    {   
        if(i==global2.length-1 && j==global2[0].length-1)
        {   
            System.out.println(ans);
            return;
        }
        else if(i==global2.length-1)
        {
            minPath(global2,i,j+1,ans+"H");
        }
        else if(j==global2[0].length-1)
        {
            minPath(global2,i+1,j,ans+"V");
        }
        else 
        {
            if(global2[i][j+1]==global2[i+1][j])
            {
                minPath(global2,i,j+1,ans+"H");
                minPath(global2,i+1,j,ans+"V");
            }
            else if(global2[i][j+1]<global2[i+1][j])
            {
                minPath(global2, i, j+1, ans+"H");
            }
            else
            {
                minPath(global2, i+1, j, ans+"V");
            }

        }                                                     
    }
    
    //REVIEWED.






    //3.GOLD_MINE Problem         (to get max-gold profit, with upward-diagonal, downward-diagonal, horizontal direction)

    public static void maxGold(int[][] goldArr,int[][] global3)             
    {
        for(int c=goldArr[0].length-1;c>=0;c--)
        {
            for(int r=0;r<goldArr.length;r++)
            {
                if(c==goldArr[0].length-1)
                {
                    global3[r][c]=goldArr[r][c];
                }
                else if(r==0)
                {
                    int max=Math.max(global3[r][c+1],global3[r+1][c+1]);
                    global3[r][c]=max+goldArr[r][c];
                }
                else if(r==goldArr.length-1)
                {
                    int max=Math.max(global3[r][c+1],global3[r-1][c+1]);
                    global3[r][c]=max+goldArr[r][c];
                }
                else
                {
                    int max1=Math.max(global3[r][c+1],global3[r+1][c+1]);
                    int max2=Math.max(max1,global3[r-1][c+1]);
                    global3[r][c]=max2+goldArr[r][c];
                }
            }
        }
    }

    //REVIEWED.






    //4. To return the count of binary strings with no two consecutive zeroes.  (without DP)
    
    public static int binaryStringCount(int n)                  
    {                                                            
        int c0=1;
        int c1=1;

        for(int i=2;i<=n;i++)
        {   
            int temp=c0;
            c0=c1;
            c1=c0+temp;
        }

        return c0+c1;
    }

    //REVIEWED.









    public static void main(String[] args)
    {
     
        //1. To get min-cost value in maze traversal.   (only two moves allowed)

        /*
        int[][] local={                                  //to get min cost in maze traversal.
            {2,7,3,8,0,3,1},
            {4,0,1,2,4,3,4},
            {3,8,5,9,0,8,1},
            {1,6,0,4,5,5,2},
            {6,2,9,5,7,0,6},
            {0,8,7,9,6,3,0},
            {1,3,5,0,5,1,0}
        };

        int[][] global=new int[local.length][local[0].length];

        System.out.println(minCost(local,global));                                   
        */

        //reviewed.

        
  




        //2. TO print min-cost path in maze traversal.         (it will use dp to get min cost,
                                                                //and directed recursion to print such paths)
        /*
        int[][] local2={                  //new maze, with multiple min-cost paths.
            {1,5,1,0,3},
            {1,4,0,2,3},
            {4,0,1,3,2},
            {2,4,0,4,1},
            {1,2,3,2,0}
        };

        int[][] global2=new int[local2.length][local2[0].length];
        
        minCost(local2,global2);                    //DP function.
        
        minPath(global2,0,0,"");                    //directed recursion function.
        */
        
        //reviewed.






        //3.Gold-mine Problem                   (directions - upper-diagonal, lower-diagonal, horizontal)
        
        /*
        int[][] goldArr={                                            //to get max_gold in maze(gold digger).
            {1,5,1,0,3},
            {1,4,0,2,3},
            {4,0,1,3,2},
            {2,4,0,4,1},
            {1,2,3,2,0}
        };

        int[][] global3=new int[goldArr.length][goldArr[0].length];
        maxGold(goldArr,global3);

        int maxProfit = Integer.MIN_VALUE;
        for(int x = 0; x < global3.length; x++) {
            if(maxProfit < global3[x][0])
                maxProfit = global3[x][0];
        }

        System.out.println("max gold profit of gold-mine is " + maxProfit);
        */

        //reviewed.






        //4. Binary-string count of length n, (with no two consicutive 0's)

        /*
        int n = 5;
        System.out.println(binaryStringCount(n));
        */

        //reviewed.




    }                    

}

File: C++_code/Lecture_28.cpp
                                        //Lecture-28
                                        //july-26

#include<iostream>
#include<vector>
#include<climits>
using namespace std;


//1.To get min-cost for 2D array traversal   (only two moves are allowed) 

int minCost(vector<vector<int>> arr) {
    vector<vector<int>> dp(arr.size(), vector<int> (arr[0].size()));
    
    for(int i = arr.size() - 1; i >= 0; i--) {
        for(int j = arr[0].size() - 1; j >= 0; j--) {
            if(i == arr.size() - 1 && j == arr[0].size() - 1) {
                dp[i][j] = arr[i][j];
            }
            else if(i == arr.size() - 1) {
                dp[i][j] = dp[i][j + 1] + arr[i][j];
            }
            else if(j == arr[0].size() - 1) {
                dp[i][j] = dp[i + 1][j] + arr[i][j];
            }
            else{
                dp[i][j] = min(dp[i + 1][j], dp[i][j + 1]) + arr[i][j];
            }
        }
    }

    return dp[0][0];
}

//REVIEWED.










//2. To get min-cost path. (using dp to get min-path, and directed recursion to get actual min cost path)

void minCostPath(vector<vector<int>> dp, int i, int j, string psf) {        //directed recursion.
    if(i == dp.size() - 1 && j == dp[0].size() - 1) {
        cout<<psf<<"\n";
        return;
    }
    
    if(i == dp.size() - 1) {
        minCostPath(dp, i, j + 1, psf + 'h');
    }
    else if(j == dp[0].size() - 1) {
        minCostPath(dp, i + 1, j, psf + 'v');
    }
    else {
        if(dp[i + 1][j] < dp[i][j + 1]) {
            minCostPath(dp, i + 1, j, psf + 'v');
        }
        else if(dp[i][j + 1] < dp[i + 1][j]) {
            minCostPath(dp, i, j + 1, psf + 'h');
        }
        else {
            minCostPath(dp, i + 1, j, psf + 'v');
            minCostPath(dp, i, j + 1, psf + 'h');
        }
    }
} 




void minCost2(vector<vector<int>> arr) {                    //(DP function)
    vector<vector<int>> dp (arr.size(), vector<int> (arr[0].size()));

    for(int i = dp.size() - 1; i >= 0; i--) {
        for(int j = dp[0].size(); j >= 0; j--) {
            if(i == dp.size() - 1 && j == dp[0].size() - 1) {
                dp[i][j] = arr[i][j];
            }
            else if(i == dp.size() - 1) {
                dp[i][j] = dp[i][j + 1] + arr[i][j];
            }
            else if(j == dp[0].size() - 1) {
                dp[i][j] = dp[i + 1][j] + arr[i][j];
            }
            else {
                dp[i][j] = min(dp[i][j + 1], dp[i + 1][j]) + arr[i][j];
            }
        }
    }

    minCostPath(dp, 0, 0, "");
}

//REVIEWED.









//3. GOLD_MINE Problem     (with upper diagonal, lower diagonal, horizontal)

int goldProfit(vector<vector<int>> goldArr) {
    vector<vector<int>> dp (goldArr.size(), vector<int> (goldArr[0].size()));

    for(int j = dp[0].size() - 1; j >= 0; j--) {
        for(int i = dp.size() - 1; i >= 0; i--) {
            if(j == dp[0].size() - 1) {
                dp[i][j] = goldArr[i][j];
            }
            else if(i == 0) {
                dp[i][j] = max(dp[i][j + 1], dp[i + 1][j + 1]) + goldArr[i][j];
            }
            else if(i == dp.size() - 1) {
                dp[i][j] = max(dp[i - 1][j + 1], dp[i][j + 1]) + goldArr[i][j];
            }
            else {
                dp[i][j] = max(dp[i - 1][j + 1], max(dp[i][j + 1], dp[i + 1][j + 1])) + goldArr[i][j];
            }
        }
    }

    int maxProfit = INT_MIN;
    for(int x = 0; x < dp.size(); x++) {
        if(maxProfit < dp[x][0])
            maxProfit = dp[x][0];
    }

    return maxProfit;
}

//REVIEWED.








//4. To count number of binary strings, without any consecutive zeroes   (implementation of fibonacci series)

int countBinaryStrings(int n) {
    int a = 1;
    int b = 1;
    int c = -1;

    for(int i = 1; i <= n; i++) {
        c = a + b;
        a = b;
        b = c;
    }

    return c;
}

//REVIEWED.









int main()
{   
    ios_base::sync_with_stdio();
    cin.tie(0);
    
    
    //1.To get min-cost value in maze traversal.     (only two moves are allowed)

    /*
    vector<vector<int>> arr {
        {2,7,3,8,0,3,1},
        {4,0,1,2,4,3,4},
        {3,8,5,9,0,8,1},
        {1,6,0,4,5,5,2},
        {6,2,9,5,7,0,6},
        {0,8,7,9,6,3,0},
        {1,3,5,0,5,1,0}
    };

    cout<<"min cost for traversal is "<<minCost(arr);
    */

   //reviewed.





   
    //2. To print min-cost path in maze traversal, (with only 2 moves)

    /*
    vector<vector<int>> arr {
        {1,5,1,0,3},
        {1,4,0,2,3},
        {4,0,1,3,2},
        {2,4,0,4,1},
        {1,2,3,2,0}
    };

    minCost2(arr);
    */

    //reviewed.







    //3.GOLD_MINE Problem        (to get max profit)
    
    /*
    vector<vector<int>> goldArr {
        {1,5,1,0,3},
        {1,4,0,2,3},
        {4,0,1,3,2},
        {2,4,0,4,1},
        {1,2,3,2,0}
    };

    cout<<"Gold-mine profit is "<<goldProfit(goldArr);
    */

    //reviewed.





    //4.Count number of binary strings without any consecutive zeroes.  (implementation of fibonacci series)

    /*
    int n = 5;
    cout<<"number of binary strings are "<<countBinaryStrings(n);
    */

    //reviewed.



}